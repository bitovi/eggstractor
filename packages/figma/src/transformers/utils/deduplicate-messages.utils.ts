export function deduplicateMessages(tokens: { warnings?: string[]; errors?: string[] }[]): {
  warnings: string[];
  errors: string[];
} {
  const warningsSet = new Set<string>();
  const errorsSet = new Set<string>();

  tokens.forEach((token) => {
    if (token.warnings) {
      token.warnings.forEach((warning) => warningsSet.add(warning));
    }
    if (token.errors) {
      token.errors.forEach((error) => errorsSet.add(error));
    }
  });

  return {
    warnings: Array.from(warningsSet),
    errors: Array.from(errorsSet),
  };
}
