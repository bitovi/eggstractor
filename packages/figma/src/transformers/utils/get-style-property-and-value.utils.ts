import { StyleToken } from '../../types';

export const getStylePropertyAndValue = (
  token: StyleToken,
): Record<string, string> => {
  const output: Record<string, string> = {
    [token.property]: token.rawValue!,
  };

  return output;
};
