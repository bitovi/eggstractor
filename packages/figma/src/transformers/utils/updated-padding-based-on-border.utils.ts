const BORDER_PROPERTIES = [
  'border',
  'border-left',
  'border-right',
  'border-top',
  'border-bottom',
  'border-width',
  'border-left-width',
  'border-right-width',
  'border-top-width',
  'border-bottom-width',
] as const;

const PADDING_PROPERTIES = [
  'padding',
  'padding-left',
  'padding-right',
  'padding-top',
  'padding-bottom',
] as const;

type CssUnit = 'px' | 'rem' | 'scss' | 'unknown' | 'none';

type ParsedCssValue =
  | {
      format: 'px' | 'rem' | 'none';
      value: number;
      text: string;
    }
  | {
      value: null;
      format: Exclude<CssUnit, 'px' | 'rem' | 'none'>;
      text: string;
    };

function parseCssValue(text: string): ParsedCssValue {
  // special case: "0" (no unit needed in CSS)
  if (text === '0' || text === '0px' || text === '0rem') {
    return { value: 0, text: '0', format: 'none' };
  }

  if (text.startsWith('$')) {
    return { value: null, text: `#{${text}}`, format: 'scss' };
  }

  // match numbers with px/rem
  const match = /^(-?\d*\.?\d+)(px|rem)?$/.exec(text);

  if (match) {
    const [, num, unit] = match;
    return {
      value: parseFloat(num),
      text,
      format: unit as 'px' | 'rem',
    };
  }

  // fallback
  return { value: null, text, format: 'unknown' };
}

/**
 * Subtracts border value from padding value, returning a valid CSS string.
 * Padding can never be negative, so the result is clamped at 0.
 */
function subtract(padding: string, border: string): string {
  const a = parseCssValue(padding);
  const b = parseCssValue(border);

  // first is zero → result is always "0"
  // second is zero → return the original first string as-is
  if (a.value === 0 || b.value === 0) {
    return a.text;
  }

  // same known units → numeric subtraction, clamped at 0
  if (
    a.format === b.format &&
    (a.format === 'px' || a.format === 'rem') &&
    (b.format === 'px' || b.format === 'rem')
  ) {
    const result = Math.max(0, a.value - b.value);
    return result === 0 ? '0' : `${result}${a.format}`;
  }

  // mismatched/unknown units → calc with clamp (no trimming)
  return `calc(${a.text} - ${b.text})`;
}

interface Padding<T = string> {
  'padding-top': T;
  'padding-right': T;
  'padding-bottom': T;
  'padding-left': T;
}

interface ParsedShorthand {
  top: string;
  right: string;
  bottom: string;
  left: string;
}

const parseShorthand = (value: string): ParsedShorthand => {
  const parts = value.trim().split(/\s+/);

  switch (parts.length) {
    case 1:
      return {
        top: parts[0],
        right: parts[0],
        bottom: parts[0],
        left: parts[0],
      };
    case 2:
      return {
        top: parts[0],
        right: parts[1],
        bottom: parts[0],
        left: parts[1],
      };
    case 3:
      return {
        top: parts[0],
        right: parts[1],
        bottom: parts[2],
        left: parts[1],
      };
    case 4:
      return {
        top: parts[0],
        right: parts[1],
        bottom: parts[2],
        left: parts[3],
      };
    default:
      throw new Error('Invalid shorthand');
  }
};

function toPaddingShorthand(paddings: Partial<Padding>): string {
  // Default missing sides to "0"
  const top = paddings['padding-top'] ?? '0';
  const right = paddings['padding-right'] ?? '0';
  const bottom = paddings['padding-bottom'] ?? '0';
  const left = paddings['padding-left'] ?? '0';

  // Case 1: all equal
  if (top === right && right === bottom && bottom === left) {
    return top;
  }

  // Case 2: top/bottom same, left/right same
  if (top === bottom && right === left) {
    return `${top} ${right}`;
  }

  // Case 3: left/right same
  if (right === left) {
    return `${top} ${right} ${bottom}`;
  }

  // Case 4: all different
  return `${top} ${right} ${bottom} ${left}`;
}

const hasBorderAndPaddingStyles = <T extends { styles: Record<string, string> }>(
  instance: T,
): boolean => {
  const hasPadding = PADDING_PROPERTIES.some((prop) => prop in instance.styles);

  if (!hasPadding) {
    return false;
  }

  return BORDER_PROPERTIES.some((prop) => prop in instance.styles);
};

const parseBorder = (borderStyleValue: string): string => {
  if (borderStyleValue === 'none') return '0';
  return borderStyleValue.split(' ')[0];
};

export const updatePaddingStylesBasedOnBorder = <
  T extends {
    styles: {
      [K in (typeof BORDER_PROPERTIES)[number] | (typeof PADDING_PROPERTIES)[number]]?: string;
    };
  },
>(
  instance: T,
): T => {
  if (!hasBorderAndPaddingStyles(instance)) {
    return instance;
  }

  const borderStyle = instance.styles['border'] ? parseBorder(instance.styles['border']) : null;
  const borderTopStyle = instance.styles['border-top']
    ? parseBorder(instance.styles['border-top'])
    : null;
  const borderRightStyle = instance.styles['border-right']
    ? parseBorder(instance.styles['border-right'])
    : null;
  const borderBottomStyle = instance.styles['border-bottom']
    ? parseBorder(instance.styles['border-bottom'])
    : null;
  const borderLeftStyle = instance.styles['border-left']
    ? parseBorder(instance.styles['border-left'])
    : null;

  const borderWidth = instance.styles['border-width']
    ? parseShorthand(instance.styles['border-width'])
    : null;

  const borders = {
    top:
      instance.styles['border-top-width'] || borderWidth?.['top'] || borderTopStyle || borderStyle,
    right:
      instance.styles['border-right-width'] ||
      borderWidth?.['right'] ||
      borderRightStyle ||
      borderStyle,
    bottom:
      instance.styles['border-bottom-width'] ||
      borderWidth?.['bottom'] ||
      borderBottomStyle ||
      borderStyle,
    left:
      instance.styles['border-left-width'] ||
      borderWidth?.['left'] ||
      borderLeftStyle ||
      borderStyle,
  };

  const paddingStyles = instance.styles['padding']
    ? parseShorthand(instance.styles['padding'])
    : null;

  const paddings: Padding<string | null> = {
    'padding-top': instance.styles['padding-top'] || paddingStyles?.['top'] || null,
    'padding-right': instance.styles['padding-right'] || paddingStyles?.['right'] || null,
    'padding-bottom': instance.styles['padding-bottom'] || paddingStyles?.['bottom'] || null,
    'padding-left': instance.styles['padding-left'] || paddingStyles?.['left'] || null,
  };

  // Only add padding properties if they exist
  const newPaddings: { [K in (typeof PADDING_PROPERTIES)[number]]?: string } = {};

  if (paddings['padding-top']) {
    newPaddings['padding-top'] = borders.top
      ? subtract(paddings['padding-top'], borders.top)
      : paddings['padding-top'];
  }

  if (paddings['padding-right']) {
    newPaddings['padding-right'] = borders.right
      ? subtract(paddings['padding-right'], borders.right)
      : paddings['padding-right'];
  }

  if (paddings['padding-bottom']) {
    newPaddings['padding-bottom'] = borders.bottom
      ? subtract(paddings['padding-bottom'], borders.bottom)
      : paddings['padding-bottom'];
  }

  if (paddings['padding-left']) {
    newPaddings['padding-left'] = borders.left
      ? subtract(paddings['padding-left'], borders.left)
      : paddings['padding-left'];
  }

  // If all individual padding properties were used, use padding shorthand as the output
  const newPaddingValues = Object.values(newPaddings);
  const allIndividualValuesUsed =
    newPaddingValues.length === 4 && newPaddingValues.every((value) => !!value);

  // Only add individual or shorthand padding properties if they exist
  const finalizedPaddingStyles: {
    [K in (typeof PADDING_PROPERTIES)[number]]?: string;
  } = {};

  if (allIndividualValuesUsed) {
    finalizedPaddingStyles['padding'] = toPaddingShorthand(newPaddings);
  } else {
    if (instance.styles['padding-top'] && newPaddings['padding-top']) {
      finalizedPaddingStyles['padding-top'] = newPaddings['padding-top'];
    }

    if (instance.styles['padding-right'] && newPaddings['padding-right']) {
      finalizedPaddingStyles['padding-right'] = newPaddings['padding-right'];
    }

    if (instance.styles['padding-bottom'] && newPaddings['padding-bottom']) {
      finalizedPaddingStyles['padding-bottom'] = newPaddings['padding-bottom'];
    }

    if (instance.styles['padding-left'] && newPaddings['padding-left']) {
      finalizedPaddingStyles['padding-left'] = newPaddings['padding-left'];
    }
  }

  // Strip out any existing padding properties to avoid conflicts
  const {
    padding: _p,
    ['padding-top']: _pt,
    ['padding-right']: _pr,
    ['padding-bottom']: _pb,
    ['padding-left']: _pl,
    ...stylesWithoutPadding
  } = instance.styles;

  return {
    ...instance,
    styles: {
      ...stylesWithoutPadding,
      // Merge in the new padding styles
      ...finalizedPaddingStyles,
    },
  };
};
