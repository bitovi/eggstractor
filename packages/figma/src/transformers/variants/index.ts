export * from './convert-to-generator-tokens';
export * from './convert-variant-group-by';
export * from './generate-combinatorial-styles';
