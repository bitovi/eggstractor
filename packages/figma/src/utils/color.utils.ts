/**
 * Converts RGBA color components to a hex string or rgba() string.
 */
export function rgbaToString(
  r: number,
  g: number,
  b: number,
  a: number,
): string {
  const rHex = Math.round(r * 255)
    .toString(16)
    .padStart(2, '0');
  const gHex = Math.round(g * 255)
    .toString(16)
    .padStart(2, '0');
  const bHex = Math.round(b * 255)
    .toString(16)
    .padStart(2, '0');

  return a === 1
    ? '#' + rHex + gHex + bHex
    : `rgba(${Math.round(r * 255)}, ${Math.round(g * 255)}, ${Math.round(b * 255)}, ${a})`;
}
