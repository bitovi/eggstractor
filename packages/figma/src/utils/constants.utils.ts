/**
 * Used for showing progress in the UI. It is capped at 95% during the time that
 * nodes are being processed to tokens. The last 5% is reserved for
 * transforming the tokens to the requested format.
 */
export const MAX_PROGRESS_PERCENTAGE = 95;
