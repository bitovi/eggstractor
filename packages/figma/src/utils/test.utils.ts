export async function serializeFigmaData(node: BaseNode): Promise<unknown> {
  // Properties to exclude from serialization
  const excludedProps = new Set([
    'inferredVariables',
    'availableInferredVariables',
    // Add other properties to exclude here
  ]);

  // Create base data with all enumerable properties
  const baseData: Record<string, unknown> = {};

  // Get all properties from the node
  for (const key in node) {
    try {
      // Skip excluded properties
      if (excludedProps.has(key)) {
        continue;
      }

      const value = node[key as keyof BaseNode];
      // Skip functions and undefined values
      if (typeof value === 'function' || value === undefined) {
        continue;
      }
      baseData[key] = value;
    } catch {
      // Some properties might throw when accessed, skip those
      continue;
    }
  }

  // Recursively process children
  if ('children' in node) {
    baseData.children = await Promise.all(
      node.children.map((child) => serializeFigmaData(child)),
    );
  }

  // Add variables collection
  const variables: Record<string, unknown> = {};

  // Helper to collect variables from node
  const collectVariables = async (node: SceneNode) => {
    if ('boundVariables' in node) {
      for (const key in node.boundVariables) {
        const vars = (
          node.boundVariables as Record<string, VariableAlias | VariableAlias[]>
        )[key];
        if (Array.isArray(vars)) {
          for (const v of vars) {
            if (v.id) {
              await collectVariableAndAliases(v.id, variables);
            }
          }
        } else if (vars?.id) {
          await collectVariableAndAliases(vars.id, variables);
        }
      }
    }

    if ('children' in node) {
      await Promise.all(
        node.children.map((child) => collectVariables(child as SceneNode)),
      );
    }
  };

  async function collectVariableAndAliases(
    variableId: string,
    variables: Record<string, unknown>,
  ) {
    const variable = await figma.variables.getVariableByIdAsync(variableId);
    if (!variable) return;

    // Store the current variable
    variables[variableId] = {
      id: variable.id,
      name: variable.name,
      resolvedType: variable.resolvedType,
      valuesByMode: variable.valuesByMode,
    };

    // Check for aliases in all modes
    for (const modeId in variable.valuesByMode) {
      const value = variable.valuesByMode[modeId];
      if (isVariableAlias(value)) {
        await collectVariableAndAliases(value.id, variables);
      }
    }
  }

  // Collect variables if it's a SceneNode
  if ('type' in node) {
    await collectVariables(node as SceneNode);
  }

  return {
    ...baseData,
    variables,
  };
}

function isVariableAlias(
  value: unknown,
): value is { type: 'VARIABLE_ALIAS'; id: string } {
  if (!value) return false;
  return (
    typeof value === 'object' &&
    'type' in value &&
    value.type === 'VARIABLE_ALIAS' &&
    'id' in value
  );
}

export async function createTestVariableResolver(testData: {
  variables: Record<string, unknown>;
}) {
  // Helper to collect and flatten all variables including aliases
  const collectAllVariables = (variables: Record<string, unknown>) => {
    const allVariables = new Map<string, unknown>();

    const addVariable = (varId: string) => {
      if (allVariables.has(varId)) return;

      const variable = variables[varId] as Variable | undefined;
      if (!variable) return;

      allVariables.set(varId, variable);

      // Recursively collect any alias references
      Object.values(variable.valuesByMode).forEach((value: unknown) => {
        if (isVariableAlias(value)) {
          addVariable(value.id);
        }
      });
    };

    // Start with root variables
    Object.keys(variables).forEach(addVariable);
    return allVariables;
  };

  const variableMap = collectAllVariables(testData.variables);

  return async (id: string): Promise<Variable | null> => {
    const variable = variableMap.get(id);
    if (!variable) return null;

    // Helper to resolve any alias values
    const resolveValue = async (value: unknown): Promise<unknown> => {
      if (isVariableAlias(value)) {
        const aliasVar = variableMap.get(value.id);
        if (!aliasVar) return null;

        const modeId = Object.keys((aliasVar as Variable).valuesByMode)[0];
        return resolveValue((aliasVar as Variable).valuesByMode[modeId]);
      }
      return value;
    };

    const modeId = Object.keys((variable as Variable).valuesByMode).sort()[0];
    const resolvedValue = await resolveValue(
      (variable as Variable).valuesByMode[modeId],
    );

    return {
      ...variable,
      resolvedType: (variable as Variable).resolvedType,
      valuesByMode: {
        [modeId]: resolvedValue,
      },
    } as Variable;
  };
}

// TODO: the types are terrible here, improve them
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function createTestData(jsonData: any) {
  const processNode = (node: BaseNode, parentNode: BaseNode | null) => {
    if (node.type === 'TEXT') {
      return {
        ...node,
        parent: parentNode,
        name: node.name,
        width: node.width || 100,
        height: node.height || 20,
        textAutoResize: node.textAutoResize || 'NONE',
      } as TextNode;
    }

    const frameNode = {
      ...node,
      type: node.type || 'FRAME',
      parent: parentNode,
      name: node.name,
      children: [],
    } as unknown as FrameNode;

    if ('children' in node) {
      // Avoid that FrameNode children is readonly with a type assertion
      (frameNode as { children: FrameNode['children'] }).children = (
        node.children as BaseNode[]
      ).map((child) => processNode(child, frameNode));
    }

    return frameNode;
  };

  const pageNode = {
    ...jsonData,
    type: 'PAGE',
    parent: null,
    children: (jsonData.children as PageNode['children']).map((child) =>
      processNode(child, null),
    ),
  } as unknown as PageNode;

  pageNode.children.forEach(
    (child) =>
      ((child as unknown as { parent: PageNode | null }).parent = pageNode),
  );

  return {
    pageNode,
    async setupTest() {
      const getVariableByIdAsync = await createTestVariableResolver(
        jsonData as Parameters<typeof createTestVariableResolver>[0],
      );
      return {
        figma: {
          currentPage: pageNode,
          root: {
            children: [pageNode], // Array of pages
          },
          loadAllPagesAsync: vi.fn().mockResolvedValue(undefined), // Mock this method
          variables: {
            getVariableByIdAsync,
          },
        },
      };
    },
  };
}
