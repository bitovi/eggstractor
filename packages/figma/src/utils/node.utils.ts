import { PathNode } from '../types';
import { sanitizeSegment } from './string.utils';

export function getParentSceneNodes(node: SceneNode): PathNode[] {
  const nodes: SceneNode[] = [];
  let current: SceneNode | null = node;

  while (current && current.parent) {
    // TODO: This is to skip the "Components" page that Figma adds automatically
    // This needs to be improved since it can be renamed
    if (current.name.toLowerCase() !== 'components') {
      nodes.unshift(current);
    }
    // TODO This doesn't properly verify that it's a SceneNode
    // we should stick to BaseNode | (BaseNode & ChildrenMixin)
    current = current.parent as SceneNode;
  }

  return nodes.map((node) => ({
    name: sanitizeSegment(node.name),
    type: node.type,
  }));
}
