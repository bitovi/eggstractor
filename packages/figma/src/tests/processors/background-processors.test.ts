import { collectTokens } from '../../services';
import { transformToCss, transformToScss } from '../../transformers';
import testData from '../fixtures/figma-test-data_background.json';
import testDataOpacity from '../fixtures/figma-test-data_opacity.json';
import { createTestData } from '../../utils';

function parseCssClass(css: string, className: string): string | null {
  const regex = new RegExp(`\\.${className}\\s*{([^}]*)}`, 'g');
  const match = regex.exec(css);
  if (!match) return null;
  return match[1].trim();
}

describe('Background Processors', () => {
  it('should process background solid correctly', async () => {
    const { setupTest } = createTestData(testData);
    const testSetup = await setupTest();

    global.figma = testSetup.figma;

    const tokens = await collectTokens(vi.fn());
    const { result: template } = transformToCss(tokens, false);

    const templateStyles = {
      solid: parseCssClass(template, 'background-solid-variable'),
      alpha: parseCssClass(template, 'background-solid-alpha-variable'),
    };

    expect(templateStyles).toMatchSnapshot('solid styles');
    expect(templateStyles.solid).toBe('background: #00464a;');
    expect(templateStyles.alpha).toBe('background: rgba(0, 70, 74, 0.5);');

    const { result: combinatorial } = transformToCss(tokens, true);

    const combinatorialStyles = {
      solid: parseCssClass(combinatorial, 'background-solid-variable'),
      alpha: parseCssClass(combinatorial, 'background-solid-alpha-variable'),
    };

    expect(combinatorialStyles).toMatchSnapshot('solid styles');
    expect(combinatorialStyles.solid).toBe('background: #00464a;');
    expect(combinatorialStyles.alpha).toBe('background: rgba(0, 70, 74, 0.5);');
  });

  it('should process background solid correctly - sass', async () => {
    const { setupTest } = createTestData(testData);
    const testSetup = await setupTest();

    global.figma = testSetup.figma;

    const tokens = await collectTokens(vi.fn());
    const { result: template } = transformToScss(tokens, false);
    expect(template).toMatchSnapshot('solid styles');
    const { result: combinatorial } = transformToScss(tokens, true);
    expect(combinatorial).toMatchSnapshot('solid styles');
  });

  it('should process opacity correctly', async () => {
    const { setupTest } = createTestData(testDataOpacity);
    const testSetup = await setupTest();

    global.figma = testSetup.figma;

    const tokens = await collectTokens(vi.fn());
    const { result: template } = transformToScss(tokens, false);
    expect(template).toMatchSnapshot('opacity styles');
    const { result: combinatorial } = transformToScss(tokens, true);
    expect(combinatorial).toMatchSnapshot('opacity styles');
  });

  it('should skip background gradient correctly', async () => {
    const { setupTest } = createTestData(testData);
    const testSetup = await setupTest();

    global.figma = testSetup.figma;

    const tokens = await collectTokens(vi.fn());
    const { result: template, warnings: templateWarnings } = transformToCss(tokens, false);

    // Test specific styles with snapshots
    const templateStyles = {
      linear: parseCssClass(template, 'background-gradient-linear-style'),
      linearAlpha: parseCssClass(template, 'background-gradient-linear-alpha-style'),
    };

    // Snapshot all styles
    expect(templateStyles).toMatchSnapshot('gradient styles');

    // We no longer warn on these values since they are skipped in the collection
    expect(templateWarnings).toHaveLength(0);

    // Keep direct assertions for critical values
    expect(templateStyles.linear).toBe(null);
    expect(templateStyles.linearAlpha).toBe(null);

    const { result: combinatorial, warnings: combinatorialWarnings } = transformToCss(tokens, true);

    // Test specific styles with snapshots
    const combinatorialStyles = {
      linear: parseCssClass(combinatorial, 'background-gradient-linear-style'),
      linearAlpha: parseCssClass(combinatorial, 'background-gradient-linear-alpha-style'),
    };

    // Snapshot all styles
    expect(combinatorialStyles).toMatchSnapshot('gradient styles');

    // We no longer warn on these values since they are skipped in the collection
    expect(combinatorialWarnings).toHaveLength(0);

    // Keep direct assertions for critical values
    expect(combinatorialStyles.linear).toBe(null);
    expect(combinatorialStyles.linearAlpha).toBe(null);
  });
});
