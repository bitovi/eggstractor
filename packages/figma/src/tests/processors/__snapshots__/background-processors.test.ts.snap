// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Background Processors > should process background solid correctly > solid styles 1`] = `
{
  "alpha": "background: rgba(0, 70, 74, 0.5);",
  "solid": "background: #00464a;",
}
`;

exports[`Background Processors > should process background solid correctly > solid styles 2`] = `
{
  "alpha": "background: rgba(0, 70, 74, 0.5);",
  "solid": "background: #00464a;",
}
`;

exports[`Background Processors > should process background solid correctly - sass > solid styles 1`] = `
"// Generated SCSS Variables
$brand-primary: #00464a;
$teal-800-50: rgba(0, 70, 74, 0.5);
$teal-800: #00464a;

// Generated SCSS Mixins
@mixin background-solid-variable {
  background: $brand-primary;
}
@mixin background-solid-alpha-variable {
  background: $teal-800-50;
}
@mixin background-solid-custom {
  background: #00464a;
}
@mixin background-solid-alpha-custom {
  background: rgba(0, 70, 74, 0.5);
}
@mixin background-solid-hex-style {
  background: #00464a;
}
@mixin background-solid-variable-style {
  background: $teal-800;
}
@mixin background-solid-alpha-hex-style {
  background: rgba(0, 70, 74, 0.5);
}
@mixin background-solid-alpha-variable-style {
  background: $teal-800-50;
}
"
`;

exports[`Background Processors > should process background solid correctly - sass > solid styles 2`] = `
"// Generated SCSS Variables
$brand-primary: #00464a;
$teal-800-50: rgba(0, 70, 74, 0.5);
$teal-800: #00464a;

// Generated SCSS Mixins
@mixin background-solid-variable {
  background: $brand-primary;
}
@mixin background-solid-alpha-variable {
  background: $teal-800-50;
}
@mixin background-solid-custom {
  background: #00464a;
}
@mixin background-solid-alpha-custom {
  background: rgba(0, 70, 74, 0.5);
}
@mixin background-solid-hex-style {
  background: #00464a;
}
@mixin background-solid-variable-style {
  background: $teal-800;
}
@mixin background-solid-alpha-hex-style {
  background: rgba(0, 70, 74, 0.5);
}
@mixin background-solid-alpha-variable-style {
  background: $teal-800-50;
}
"
`;

exports[`Background Processors > should process opacity correctly > opacity styles 1`] = `
"// Generated SCSS Variables
$brand-primary: #00464a;
$opacity-48: 0.48;
$corner-radius-primary: 1rem;
$size-16: 1rem;
$font-weight-normal: 31.25rem;
$text-primary: #a3adad;
$size-48: 0.48;
$spacing-9: 0.48;

// Generated SCSS Mixins
@mixin opacity-custom {
  background: $brand-primary;
  opacity: 0.5;
}
@mixin opacity-variable {
  background: $brand-primary;
  opacity: $opacity-48;
}
@mixin opacity-custom-group {
  padding: 0rem;
  background: #ffffff;
  opacity: 0.5;
}
@mixin opacity-custom-group-opacity-custom {
  background: #00464a;
  opacity: 0.5;
}
@mixin opacity-custom-group-lorem-ipsum {
  display: flex;
  flex-direction: column;
  color: #ffffff;
  font-family: Inter;
  font-size: $corner-radius-primary;
  font-weight: 500;
  font-style: normal;
  line-height: $size-16;
  width: 4.25rem;
  height: 1rem;
}
@mixin opacity-variable-group {
  padding: 0rem;
  background: $text-primary;
  opacity: $size-48;
}
@mixin opacity-variable-group-opacity-custom {
  background: $brand-primary;
  opacity: $size-48;
}
@mixin opacity-variable-group-lorem-ipsum {
  display: flex;
  flex-direction: column;
  color: $text-primary;
  font-family: Inter;
  font-size: $corner-radius-primary;
  font-weight: 500;
  font-style: normal;
  line-height: $size-16;
  width: 4.25rem;
  height: 1rem;
}
"
`;

exports[`Background Processors > should process opacity correctly > opacity styles 2`] = `
"// Generated SCSS Variables
$brand-primary: #00464a;
$opacity-48: 0.48;
$corner-radius-primary: 1rem;
$size-16: 1rem;
$font-weight-normal: 31.25rem;
$text-primary: #a3adad;
$size-48: 0.48;
$spacing-9: 0.48;

// Generated SCSS Mixins
@mixin opacity-custom {
  background: $brand-primary;
  opacity: 0.5;
}
@mixin opacity-variable {
  background: $brand-primary;
  opacity: $opacity-48;
}
@mixin opacity-custom-group {
  padding: 0rem;
  background: #ffffff;
  opacity: 0.5;
}
@mixin opacity-custom-group-opacity-custom {
  background: #00464a;
  opacity: 0.5;
}
@mixin opacity-custom-group-lorem-ipsum {
  display: flex;
  flex-direction: column;
  color: #ffffff;
  font-family: Inter;
  font-size: $corner-radius-primary;
  font-weight: 500;
  font-style: normal;
  line-height: $size-16;
  width: 4.25rem;
  height: 1rem;
}
@mixin opacity-variable-group {
  padding: 0rem;
  background: $text-primary;
  opacity: $size-48;
}
@mixin opacity-variable-group-opacity-custom {
  background: $brand-primary;
  opacity: $size-48;
}
@mixin opacity-variable-group-lorem-ipsum {
  display: flex;
  flex-direction: column;
  color: $text-primary;
  font-family: Inter;
  font-size: $corner-radius-primary;
  font-weight: 500;
  font-style: normal;
  line-height: $size-16;
  width: 4.25rem;
  height: 1rem;
}
"
`;

exports[`Background Processors > should skip background gradient correctly > gradient styles 1`] = `
{
  "linear": null,
  "linearAlpha": null,
}
`;

exports[`Background Processors > should skip background gradient correctly > gradient styles 2`] = `
{
  "linear": null,
  "linearAlpha": null,
}
`;
